source/structured_data.cc

```diff
@@ -1042,40 +995,154 @@ namespace aspect
 
 
 
+    namespace
+    {
+      /**
+       * Determines which of the dimensions of a position are aligned with
+       * a certain @p boundary_id. E.g. the left boundary of a box
+       * model extents in the y and z direction (position[1] and
+       * position[2]), therefore the function would return [1,2] for dim==3
+       * or [1] for dim==2. We are lucky that these indices are identical
+       * for all existing geometries (if we use natural coordinates),
+       * therefore we do not need to distinguish between them. If we
+       * introduce a geometry model for which boundaries are not aligned
+       * with natural coordinates this needs to become a function in
+       * the interface of the geometry model and needs to be specialized
+       * for distinct geometry models.
+       */
+      template <int dim>
+      std::array<unsigned int,dim-1>
+      get_boundary_dimensions (const types::boundary_id boundary_id)
+      {
+        std::array<unsigned int,dim-1> boundary_dimensions;
+
+        switch (dim)
+          {
+            case 2:
+              if ((boundary_id == 2) || (boundary_id == 3) || (boundary_id == 4) || (boundary_id == 5))
+                {
+                  boundary_dimensions[0] = 0;
+                }
+              else if ((boundary_id == 0) || (boundary_id == 1))
+                {
+                  boundary_dimensions[0] = 1;
+                }
+              else
+                {
+                  boundary_dimensions[0] = numbers::invalid_unsigned_int;
+                  AssertThrow(false,ExcNotImplemented());
+                }
+              break;
+
+            case 3:
+              if ((boundary_id == 4) || (boundary_id == 5) || (boundary_id == 8) || (boundary_id == 9))
+                {
+                  boundary_dimensions[0] = 0;
+                  boundary_dimensions[1] = 1;
+                }
+              else if ((boundary_id == 0) || (boundary_id == 1))
+                {
+                  boundary_dimensions[0] = 1;
+                  boundary_dimensions[1] = 2;
+                }
+              else if ((boundary_id == 2) || (boundary_id == 3) || (boundary_id == 6) || (boundary_id == 7))
+                {
+                  boundary_dimensions[0] = 0;
+                  boundary_dimensions[1] = 2;
+                }
+              else
+                {
+                  boundary_dimensions[0] = numbers::invalid_unsigned_int;
+                  boundary_dimensions[1] = numbers::invalid_unsigned_int;
+                  AssertThrow(false,ExcNotImplemented());
+                }
+
+              break;
+
+            default:
+              for (unsigned int d=0; d<dim-1; ++d)
+                boundary_dimensions[d] = numbers::invalid_unsigned_int;
```

Comment:
I think you can use ``boundary_dimensions.fill(numbers::invalid_unsigned_int);``